src/io/gff.rs

```diff
@@ -0,0 +1,305 @@
+use std::convert::AsRef;
+use std::collections::HashMap;
+
+use csv;
+
+use io::Strand;
+
+/// A GFF reader.
+pub struct Reader<R: io::Read> {
+    inner: csv::Reader<R>,
+}
+
+impl Reader<fs::File> {
+    /// Read GFF from given file path.
+    pub fn from_file<P: AsRef<Path>>(path: P) -> io::Result<Self> {
+        fs::File::open(path).map(Reader::new)
+    }
+}
+
+
+impl<R: io::Read> Reader<R> {
+    /// Create a new GFF reader given an instance of `io::Read`.
+    pub fn new(reader: R) -> Self {
+        Reader {
+            inner: csv::Reader::from_reader(reader).delimiter(b'\t').has_headers(false)
+        }
+    }
+
+    /// Iterate over all records.
+    pub fn records(&mut self) -> Records<R> {
+        Records { inner: self.inner.decode() }
+    }
+}
+
+/// A GFF record.
+pub struct Records<'a, R: 'a + io::Read> {
+    inner: csv::DecodedRecords<'a, R, (String, String, String, u64, u64, String, String, String, String)>,
+}
+
+
+impl<'a, R: io::Read> Iterator for Records<'a, R> {
+    type Item = csv::Result<Record>;
+
+    fn next(&mut self) -> Option<csv::Result<Record>> {
+        self.inner.next().map(|res| {
+            res.map(|(seqname, source, feature_type, start, end, score, strand, frame, attributes)| {
+                Record {
+                    seqname: seqname,
+                    source: source,
+                    feature_type: feature_type,
+                    start: start,
+                    end: end,
+                    score: score,
+                    strand: strand,
+                    frame: frame,
+                    attributes: csv::Reader::from_string(attributes)
+                        .delimiter(b'=')
+                        .record_terminator(csv::RecordTerminator::Any(b';'))
+                        .has_headers(false)
+                        .decode().collect::<csv::Result<HashMap<String, String>>>().unwrap(),
+                }
+            })
+        })
+    }
+}
+
+
+/// A GFF writer.
+pub struct Writer<W: io::Write> {
+    inner: csv::Writer<W>,
+}
+
+
+impl Writer<fs::File> {
+    /// Write to a given file path.
+    pub fn to_file<P: AsRef<Path>>(path: P) -> io::Result<Self> {
+        fs::File::create(path).map(Writer::new)
+    }
+}
+
+
+impl<W: io::Write> Writer<W> {
+    /// Write to a given writer.
+    pub fn new(writer: W) -> Self {
+        Writer { inner: csv::Writer::from_writer(writer).delimiter(b'\t').flexible(true) }
+    }
+
+    /// Write a given GFF record.
+    pub fn write(&mut self, record: Record) -> csv::Result<()> {
+        if !record.attributes.is_empty() {
+            self.inner.encode((
+                record.seqname,
+                record.source,
+                record.feature_type,
+                record.start,
+                record.end,
+                record.score,
+                record.strand,
+                record.frame,
+                record.attributes.iter().map(|(a, b)| format!("{}={}", a, b)).collect::<Vec<_>>().join(";"),
```

Comment:
Can you move this to the top of the function and assign it to a variable? Then you don't have to duplicate the `self.inner.encode`.